lib/routes/account.js

```diff
@@ -864,6 +871,64 @@ module.exports = function (
       }
     },
     {
+      method: 'POST',
+      path: '/account/devices/notify',
+      config: {
+        auth: {
+          strategy: 'sessionTokenWithDevice'
+        },
+        validate: {
+          payload: isA.alternatives().try(
+            isA.object({
+              to: isA.string().valid('all').required(),
+              excluded: isA.array().items(isA.string().length(32).regex(HEX_STRING)).optional(),
+              payload: isA.object().required(),
+              TTL: isA.number().integer().min(0).optional()
+            }),
+            isA.object({
+              to: isA.array().items(isA.string().length(32).regex(HEX_STRING)).required(),
+              payload: isA.object().required(),
+              TTL: isA.number().integer().min(0).optional()
+            })
+          )
+        },
+        response: {
+          schema: {}
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Account.devicesNotify', request)
+        var body = request.payload
+        var sessionToken = request.auth.credentials
+        var uid = sessionToken.uid
+        var payload = body.payload
+
+        try {
+          if (!validatePushPayload(payload)) {
+            throw error.invalidRequestParameter('invalid payload')
+          }
+          var pushOptions = {
+            data: payload
+          }
+          if (body.excluded) {
+            pushOptions.excludedDeviceIds = body.excluded
+          }
+          if (body.TTL) {
+            pushOptions.TTL = body.TTL
+          }
+          if (body.to === 'all') {
+            push.pushToAllDevices(uid, 'devicesNotify', pushOptions)
+          } else {
+            push.pushToDevices(uid, body.to, 'devicesNotify', pushOptions)
+          }
+          reply({})
+        }
+        catch (e) {
+          reply(e)
+        }
```

Comment:
@eoger do we have use the same pattern anywhere else? if not could this be changed so that we do not call `reply` twice if the first reply fails for whatever reason. 

If something like this happens (except with an Error inside of `reply`) then it will call reply twice. 
![](http://v14d.com/i/579249a4d748a.jpg)